cloudsync/smartsync.py

```diff
@@ -0,0 +1,308 @@
+import time
+import logging
+from dataclasses import dataclass
+from typing import Optional, Tuple, TYPE_CHECKING, Callable, List, Set
+from cloudsync import CloudSync, SyncManager, SyncState, SyncEntry
+from cloudsync.types import LOCAL, REMOTE, DIRECTORY, OInfo, DirInfo
+log = logging.getLogger(__name__)
+
+if TYPE_CHECKING:  # pragma: no cover
+    from .provider import Provider
+    from cloudsync import Storage
+
+class SmartSyncManager(SyncManager):
+    """Class to allow for syncing files only on demand."""
+    def pre_sync(self, sync: SyncEntry) -> bool:  # pylint: disable=too-many-branches
+        finished = super().pre_sync(sync)
+        if not finished:
+            finished = not (sync in self.state.requestset or sync[REMOTE].otype == DIRECTORY)
+        return finished
+
+    def get_parent_conflicts(self, sync: SyncEntry, changed) -> List[SyncEntry]:
+        """Returns list of parent conflicts."""
+        done = False
+        pcs = []
+        new_pc = sync
+        while not done:
+            pc = self._get_parent_conflict(new_pc, changed)
+            # if pc:  # TODO: test if this is necessary
+            #     pc.get_latest()
+            if pc and pc[REMOTE].path != new_pc[REMOTE].path:
+                pcs.append(pc)
+                new_pc = pc
+            else:
+                done = True
+        pcs.reverse()
+        return pcs
+
+
+class SmartSyncState(SyncState):
+    """Enhances the syncstate to support smart syncing."""
+    def __init__(self,
+                 providers: Tuple['Provider', 'Provider'],
+                 storage: Optional['Storage'] = None,
+                 tag: Optional[str] = None,
+                 shuffle: bool = False,
+                 prioritize: Callable[[int, str], int] = None):
+        self.requestset: Set[SyncEntry] = set()
+        self.excludeset: Set[SyncEntry] = set()
+        self._callbacks: List[callable] = list()
+        super().__init__(providers, storage, tag, shuffle, prioritize)
+
+    def register_auto_sync_callback(self, callback):
```

Comment:
Might make sense to move this up a level and expose this interface at the `SmartCloudSync` level

Reply:
done